File: clients/android/app/src/main/java/com/nabto/tmuxremote/services/CBORHelpers.kt
package com.nabto.tmuxremote.services

import com.fasterxml.jackson.dataformat.cbor.CBORFactory
import com.fasterxml.jackson.dataformat.cbor.CBORGenerator
import com.fasterxml.jackson.dataformat.cbor.CBORParser
import com.nabto.tmuxremote.models.SessionInfo
import com.nabto.tmuxremote.patterns.PatternMatch
import com.nabto.tmuxremote.patterns.PatternType
import com.nabto.tmuxremote.patterns.ResolvedAction
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

sealed class ControlStreamEvent {
    data class Sessions(val sessions: List<SessionInfo>) : ControlStreamEvent()
    data class PatternPresent(val match: PatternMatch) : ControlStreamEvent()
    data class PatternUpdate(val match: PatternMatch) : ControlStreamEvent()
    data class PatternGone(val instanceId: String) : ControlStreamEvent()
}

object CBORHelpers {

    private val factory = CBORFactory()

    // MARK: - Encoding

    fun encodeAttach(session: String, cols: Int, rows: Int): ByteArray {
        return encodeMap(
            "session" to session,
            "cols" to cols,
            "rows" to rows
        )
    }

    fun encodeCreate(session: String, cols: Int, rows: Int, command: String? = null): ByteArray {
        val pairs = mutableListOf<Pair<String, Any>>(
            "session" to session,
            "cols" to cols,
            "rows" to rows
        )
        if (command != null) {
            pairs.add("command" to command)
        }
        return encodeMap(*pairs.toTypedArray())
    }

    fun encodeResize(cols: Int, rows: Int): ByteArray {
        return encodeMap(
            "cols" to cols,
            "rows" to rows
        )
    }

    fun encodePatternResolve(instanceId: String, decision: String, keys: String? = null): ByteArray {
        val pairs = mutableListOf<Pair<String, Any>>(
            "type" to "pattern_resolve",
            "instance_id" to instanceId,
            "decision" to decision
        )
        if (keys != null) {
            pairs.add("keys" to keys)
        }
        val payload = encodeMap(*pairs.toTypedArray())
        val lengthPrefix = ByteBuffer.allocate(4).putInt(payload.size).array()
        return lengthPrefix + payload
    }

    fun encodePairingPayload(username: String): ByteArray {
        return encodeMap("Username" to username)
    }

    private fun encodeMap(vararg pairs: Pair<String, Any>): ByteArray {
        val baos = ByteArrayOutputStream()
        val gen: CBORGenerator = factory.createGenerator(baos)
        gen.writeStartObject()
        for ((key, value) in pairs) {
            gen.writeFieldName(key)
            when (value) {
                is String -> gen.writeString(value)
                is Int -> gen.writeNumber(value)
                is Long -> gen.writeNumber(value)
                else -> gen.writeString(value.toString())
            }
        }
        gen.writeEndObject()
        gen.close()
        return baos.toByteArray()
    }

    // MARK: - Decoding

    fun decodeSessions(data: ByteArray): List<SessionInfo> {
        val parser = factory.createParser(data)
        return decodeSessionsFromParser(parser)
    }

    fun decodeControlStreamEvent(data: ByteArray): ControlStreamEvent? {
        val map = decodeMap(data) ?: return null
        val type = map["type"] as? String ?: return null

        return when (type) {
            "sessions" -> {
                @Suppress("UNCHECKED_CAST")
                val sessionList = map["sessions"] as? List<Map<String, Any>> ?: return null
                val sessions = sessionList.mapNotNull { decodeSessionMap(it) }
                ControlStreamEvent.Sessions(sessions)
            }
            "pattern_present" -> {
                val match = decodePatternInstance(map) ?: return null
                ControlStreamEvent.PatternPresent(match)
            }
            "pattern_update" -> {
                val match = decodePatternInstance(map) ?: return null
                ControlStreamEvent.PatternUpdate(match)
            }
            "pattern_gone" -> {
                val instanceId = map["instance_id"] as? String ?: return null
                ControlStreamEvent.PatternGone(instanceId)
            }
            else -> null
        }
    }

    private fun decodePatternInstance(map: Map<String, Any>): PatternMatch? {
        val instanceId = map["instance_id"] as? String ?: return null
        val patternId = map["pattern_id"] as? String ?: return null
        val patternTypeStr = map["pattern_type"] as? String ?: return null
        val patternType = PatternType.fromString(patternTypeStr) ?: return null
        val prompt = map["prompt"] as? String

        @Suppress("UNCHECKED_CAST")
        val actionsList = map["actions"] as? List<Map<String, Any>> ?: return null
        val actions = actionsList.mapNotNull { actionMap ->
            val label = actionMap["label"] as? String ?: return@mapNotNull null
            val keys = actionMap["keys"] as? String ?: return@mapNotNull null
            ResolvedAction(label = label, keys = keys)
        }
        if (actions.isEmpty()) return null

        val revision = when (val rev = map["revision"]) {
            is Int -> rev
            is Long -> rev.toInt()
            else -> 1
        }

        return PatternMatch(
            id = instanceId,
            patternId = patternId,
            patternType = patternType,
            prompt = prompt,
            actions = actions,
            revision = revision
        )
    }

    private fun decodeSessionsFromParser(parser: CBORParser): List<SessionInfo> {
        val sessions = mutableListOf<SessionInfo>()
        // Expect array start
        if (parser.nextToken() == null) return emptyList()
        // Could be an array of session objects
        if (!parser.isExpectedStartArrayToken) return emptyList()
        while (parser.nextToken() != null && !parser.currentToken.isStructEnd) {
            val sessionMap = parseObject(parser)
            decodeSessionMap(sessionMap)?.let { sessions.add(it) }
        }
        parser.close()
        return sessions
    }

    private fun decodeSessionMap(map: Map<String, Any>): SessionInfo? {
        val name = map["name"] as? String ?: return null
        val cols = (map["cols"] as? Number)?.toInt() ?: 0
        val rows = (map["rows"] as? Number)?.toInt() ?: 0
        val attached = (map["attached"] as? Number)?.toInt() ?: 0
        return SessionInfo(name = name, cols = cols, rows = rows, attached = attached)
    }

    private fun decodeMap(data: ByteArray): Map<String, Any>? {
        return try {
            val parser = factory.createParser(data)
            parser.nextToken() // START_OBJECT
            if (!parser.isExpectedStartObjectToken) {
                parser.close()
                return null
            }
            val result = parseObject(parser)
            parser.close()
            result
        } catch (_: Exception) {
            null
        }
    }

    private fun parseObject(parser: CBORParser): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        while (parser.nextToken() != null && !parser.currentToken.isStructEnd) {
            val key = parser.currentName ?: continue
            parser.nextToken()
            val value = parseValue(parser)
            if (value != null) {
                map[key] = value
            }
        }
        return map
    }

    private fun parseValue(parser: CBORParser): Any? {
        return when {
            parser.currentToken.isNumeric -> parser.longValue
            parser.currentToken.isBoolean -> parser.booleanValue
            parser.isExpectedStartObjectToken -> parseObject(parser)
            parser.isExpectedStartArrayToken -> parseArray(parser)
            parser.currentToken == com.fasterxml.jackson.core.JsonToken.VALUE_STRING ->
                parser.text
            parser.currentToken == com.fasterxml.jackson.core.JsonToken.VALUE_NULL -> null
            else -> parser.text
        }
    }

    private fun parseArray(parser: CBORParser): List<Any?> {
        val list = mutableListOf<Any?>()
        while (parser.nextToken() != null && !parser.currentToken.isStructEnd) {
            list.add(parseValue(parser))
        }
        return list
    }
}

File: clients/ios/TmuxRemote/Services/CBORHelpers.swift
import Foundation
import SwiftCBOR

/// Decoded control stream event from agent.
enum ControlStreamEvent {
    case sessions([SessionInfo])
    case patternPresent(PatternMatch)
    case patternUpdate(PatternMatch)
    case patternGone(String)
}

enum CBORHelpers {

    // MARK: - Encoding

    static func encodeAttach(session: String, cols: Int, rows: Int) -> Data {
        let cbor: CBOR = .map([
            .utf8String("session"): .utf8String(session),
            .utf8String("cols"): .unsignedInt(UInt64(cols)),
            .utf8String("rows"): .unsignedInt(UInt64(rows))
        ])
        return Data(cbor.encode())
    }

    static func encodeCreate(session: String, cols: Int, rows: Int, command: String? = nil) -> Data {
        var map: [CBOR: CBOR] = [
            .utf8String("session"): .utf8String(session),
            .utf8String("cols"): .unsignedInt(UInt64(cols)),
            .utf8String("rows"): .unsignedInt(UInt64(rows))
        ]
        if let command = command {
            map[.utf8String("command")] = .utf8String(command)
        }
        let cbor: CBOR = .map(map)
        return Data(cbor.encode())
    }

    static func encodeResize(cols: Int, rows: Int) -> Data {
        let cbor: CBOR = .map([
            .utf8String("cols"): .unsignedInt(UInt64(cols)),
            .utf8String("rows"): .unsignedInt(UInt64(rows))
        ])
        return Data(cbor.encode())
    }

    static func encodePatternResolve(instanceId: String, decision: String, keys: String?) -> Data {
        var map: [CBOR: CBOR] = [
            .utf8String("type"): .utf8String("pattern_resolve"),
            .utf8String("instance_id"): .utf8String(instanceId),
            .utf8String("decision"): .utf8String(decision)
        ]
        if let keys {
            map[.utf8String("keys")] = .utf8String(keys)
        }

        let payload = Data(CBOR.map(map).encode())
        var lengthPrefix = Data(count: 4)
        let len = UInt32(payload.count).bigEndian
        withUnsafeBytes(of: len) { lengthPrefix.replaceSubrange(0..<4, with: $0) }
        return lengthPrefix + payload
    }

    // MARK: - Decoding

    static func decodeControlMessage(from data: Data) -> [SessionInfo] {
        guard let decoded = try? CBOR.decode([UInt8](data)) else { return [] }
        guard case .map(let outerMap) = decoded else { return [] }
        guard case .utf8String("sessions") = outerMap[.utf8String("type")] else { return [] }
        guard case .array(let items) = outerMap[.utf8String("sessions")] else { return [] }
        return decodeSessionArray(items)
    }

    static func decodeControlStreamEvent(from data: Data) -> ControlStreamEvent? {
        guard let decoded = try? CBOR.decode([UInt8](data)) else { return nil }
        guard case .map(let outerMap) = decoded else { return nil }

        guard case .utf8String(let typeStr) = outerMap[.utf8String("type")] else {
            return nil
        }

        switch typeStr {
        case "sessions":
            if case .array(let items) = outerMap[.utf8String("sessions")] {
                return .sessions(decodeSessionArray(items))
            }
            return nil
        case "pattern_present":
            guard let match = decodePatternInstance(outerMap) else { return nil }
            return .patternPresent(match)
        case "pattern_update":
            guard let match = decodePatternInstance(outerMap) else { return nil }
            return .patternUpdate(match)
        case "pattern_gone":
            guard case .utf8String(let instanceId) = outerMap[.utf8String("instance_id")] else {
                return nil
            }
            return .patternGone(instanceId)
        default:
            return nil
        }
    }

    private static func decodePatternType(_ value: CBOR?) -> PatternType? {
        guard case .utf8String(let typeStr) = value else { return nil }
        switch typeStr {
        case "yes_no": return .yesNo
        case "numbered_menu": return .numberedMenu
        case "accept_reject": return .acceptReject
        default: return nil
        }
    }

    private static func decodePatternInstance(_ map: [CBOR: CBOR]) -> PatternMatch? {
        guard case .utf8String(let instanceId) = map[.utf8String("instance_id")] else { return nil }
        guard case .utf8String(let patternId) = map[.utf8String("pattern_id")] else { return nil }
        guard let patternType = decodePatternType(map[.utf8String("pattern_type")]) else { return nil }

        let prompt: String?
        if case .utf8String(let p) = map[.utf8String("prompt")] {
            prompt = p
        } else {
            prompt = nil
        }

        var actions: [ResolvedAction] = []
        if case .array(let items) = map[.utf8String("actions")] {
            for item in items {
                guard case .map(let actionMap) = item else { continue }
                guard case .utf8String(let label) = actionMap[.utf8String("label")] else { continue }
                guard case .utf8String(let keys) = actionMap[.utf8String("keys")] else { continue }
                actions.append(ResolvedAction(label: label, keys: keys))
            }
        }

        let revision: Int
        if case .unsignedInt(let value) = map[.utf8String("revision")] {
            revision = Int(value)
        } else {
            revision = 1
        }

        guard !actions.isEmpty else { return nil }

        return PatternMatch(
            id: instanceId,
            patternId: patternId,
            patternType: patternType,
            prompt: prompt,
            actions: actions,
            revision: revision
        )
    }

    struct AgentStatus {
        let version: String
        let friendlyName: String?
    }

    static func decodeStatus(from data: Data) -> AgentStatus? {
        guard let decoded = try? CBOR.decode([UInt8](data)) else { return nil }
        guard case .map(let map) = decoded else { return nil }

        let version: String
        if case .utf8String(let v) = map[.utf8String("version")] {
            version = v
        } else {
            return nil
        }

        let friendlyName: String?
        if case .utf8String(let n) = map[.utf8String("friendly_name")] {
            friendlyName = n
        } else {
            friendlyName = nil
        }

        return AgentStatus(version: version, friendlyName: friendlyName)
    }

    static func decodeSessions(from data: Data) -> [SessionInfo] {
        guard let decoded = try? CBOR.decode([UInt8](data)) else { return [] }
        guard case .array(let items) = decoded else { return [] }
        return decodeSessionArray(items)
    }

    private static func decodeSessionArray(_ items: [CBOR]) -> [SessionInfo] {
        var sessions: [SessionInfo] = []
        for item in items {
            guard case .map(let map) = item else { continue }

            let name: String
            if case .utf8String(let s) = map[.utf8String("name")] {
                name = s
            } else {
                continue
            }

            let cols: Int
            if case .unsignedInt(let v) = map[.utf8String("cols")] {
                cols = Int(v)
            } else {
                cols = 0
            }

            let rows: Int
            if case .unsignedInt(let v) = map[.utf8String("rows")] {
                rows = Int(v)
            } else {
                rows = 0
            }

            let attached: Int
            if case .unsignedInt(let v) = map[.utf8String("attached")] {
                attached = Int(v)
            } else {
                attached = 0
            }

            sessions.append(SessionInfo(name: name, cols: cols, rows: rows, attached: attached))
        }
        return sessions
    }
}
